### PR TITLE
[Bugfix] Fix serviceexport deletion leaks targetgroup

### PR DESCRIPTION
**Pull request by zijun726911**
**What type of PR is this?**
bug fix for serviceexport deletion leaks targetgroup issue  #269 

**Changes**:

- Refactor (t *targetGroupModelBuildTask) BuildTargetGroup(), make sure it can build model for serviceExport deletion request (the key point is the built model should have `IsDeleted: true`,`tg.Spec.LatticeID = dsTG.ID` )
- Split the ServiceExport creation and delete logic into 2 smaller helper functions: `buildTargetGroupForServiceExportCreation` and `buildTargetGroupForServiceExportDeletion` to make code eaiser to read
- Make sure we add/delete/update the correct latticeDataStore TargetGroup cache, so that the garbage collector logic: `SynthesizeSDKTargetGroups()` could be able to identifed the correct stale tgs, check the code changes for detail

**Testing done on this change**:

Run multiple times, all e2etest cases pass
```
[AfterSuite] PASSED [30.210 seconds]
------------------------------

Ran 13 of 13 Specs in 2320.742 seconds
SUCCESS! -- 13 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestIntegration (2321.24s)
```
Checked the aws vpc lattice console after e2etest finished, no target group leaking

### Manual test:

- Scenario 1: stand-alone serviceExport(No httproute created, in other words, created targetGroups don't have latticeServiceRule refer to it):

    - Scenario 1.1 create k8Service, serviceExport,tg created success, delete k8Service first, then delete serviceExport, tg, targets deleted success, no TG leak
    - Scenario 1.2 create k8Service, serviceExport, delete serviceExport first, then delete k8Service, , tg, targets deleted success, no TG leak

- Scenario 2: serviceExport's serviceImport referenced by a httproute
    - Scenario 2.1 Create httproute, k8Service, serviceImport serviceExport, all lattice created success. then, delete serviceExport first, tg will not be deleted as expected, since this tg is refered by latticeService. `kubectl delete serviceExport inventory-ver2` will not be stuck. Then, delete the httproute, the garbage collector logic `SynthesizeSDKTargetGroups()` be able to clean this inventory-ver2 tg, no any lattice resource leak
    - Scenario 2.2 Create httproute, k8Service, serviceImport serviceExport, all lattice created success. then, delete the the httproute first, then the  serviceExport become "stand-alone" status, delete this serviceExport could delete `inventory-ver2` tg success

**Will this PR introduce any new dependencies?**: No

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**: No

**Does this PR introduce any user-facing change?**: No

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/latticestore/latticestore.go [hunk 1]
```diff
@@ -392,6 +402,11 @@ func (ds *LatticeDataStore) SetTargetGroupByBackendRef(name string, routeName st
 
 }
 
+// TODO: make this DelTargetGroup() method to be private, create seperate methods:
```
Comment (xWink) [inline pkg/latticestore/latticestore.go:405]: What's the purpose of these TODO items?

**Review by xWink — COMMENTED** [review]
I don't see any issues, but I'm still learning about this logic flow. Would like to see other reviewers comment before approving. Note to Mikhail if he reviews, don't worry about glog usage, I will be fixing it as part of my PR anyways.

File: pkg/gateway/model_build_targetgroup_test.go [hunk 2]
```diff
@@ -145,7 +145,10 @@ func Test_TGModelByServicexportBuild(t *testing.T) {
 			}
 
 			ds := latticestore.NewLatticeDataStore()
-
+			if !tt.svcExport.DeletionTimestamp.IsZero() {
+				tgName := latticestore.TargetGroupName(tt.svcExport.Name, tt.svcExport.Namespace)
+				ds.AddTargetGroup(tgName, "vpc-123456789", "123456789", "tg-123", false, "")
+			}
```
Comment (zijun726911) [inline pkg/gateway/model_build_targetgroup_test.go:151]: when test the ServiceExport deletion, the latticestore should already have a entry for it, so that the built tg.Spec.LatticeID will have this tgId, and it will be pass to the target_group_manage to delte this tg

File: pkg/gateway/model_build_targetgroup.go [hunk 3]
```diff
@@ -172,65 +164,82 @@ func (t *latticeServiceModelBuildTask) buildTargets(ctx context.Context) error {
...
-			K8SServiceNamespace: t.serviceExport.Namespace,
-			Protocol:            "HTTP",
-			ProtocolVersion:     vpclattice.TargetGroupProtocolVersionHttp1,
+			Port:                  80,
+			IsServiceImport:       false,
+			IsServiceExport:       true,
+			K8SServiceName:        t.serviceExport.Name,
+			K8SServiceNamespace:   t.serviceExport.Namespace,
+			K8SHTTPRouteName:      "", // When build TG model for serviceExport, the httproutename should be ""
+			K8SHTTPRouteNamespace: "",
```
Comment (zijun726911) [inline pkg/gateway/model_build_targetgroup.go:207]: Explicitly show that  When build TG model for serviceExport, the httproutename should be "",  this is imporment, if we are not aware that it can cause some bug that hard to figure out. I will try to imporve our code to remove such "implicit convention"

File: pkg/gateway/model_build_targetgroup.go [hunk 4]
```diff
@@ -72,47 +73,38 @@ func (b *targetGroupBuilder) Build(ctx context.Context, srvExport *mcs_api.Servi
...
+	if err := task.buildModel(ctx); err != nil {
 		return task.stack, task.targetGroup, corev1.ErrIntOverflowGenerated
 	}
 
 	return task.stack, task.targetGroup, nil
 }
 
-// for serviceexport
-func (t *targetGroupModelBuildTask) run(ctx context.Context) error {
-	/*
```
Comment (zijun726911) [inline pkg/gateway/model_build_targetgroup.go:84]: Removed this small wrapper function `run()`

File: pkg/latticestore/latticestore.go (hunk 1 above)
Comment (zijun726911) [reply pkg/latticestore/latticestore.go:405]: Current we use this way to delete a tg cache record that this k8sSerivce backendref in a httproute: 			

`t.latticeDataStore.DelTargetGroup(resTargetGroup.Spec.Name, resTargetGroup.Spec.Config.K8SHTTPRouteName, false)`


we use this way to delete a tg cache record for which this serviceImport backendref in a httproute: 
`t.latticeDataStore.DelTargetGroup(resTargetGroup.Spec.Name, "", true)`		


And we use this way to delete a tg cache record for serviceExport: 
`t.latticeDataStore.DelTargetGroup(resTargetGroup.Spec.Name, "", false)`		


I think these functions are really confusing, we should improve them

File: pkg/deploy/lattice/target_group_synthesizer.go [hunk 5]
```diff
@@ -423,7 +422,7 @@ func (t *targetGroupSynthesizer) SynthesizeTriggeredTargetGroupsDeletion(ctx con
 
 		if resTargetGroup.Spec.Config.IsServiceImport {
 			glog.V(2).Infof("Deleting service import target group from local datastore %v", resTargetGroup.Spec.LatticeID)
-			t.latticeDataStore.DelTargetGroup(resTargetGroup.Spec.Name, resTargetGroup.Spec.Config.K8SHTTPRouteName, resTargetGroup.Spec.Config.IsServiceImport)
+			t.latticeDataStore.DelTargetGroup(resTargetGroup.Spec.Name, "", resTargetGroup.Spec.Config.IsServiceImport)
 		} else {
```
[the remark this answers is not in the document]
Comment (mikhail-aws) [reply pkg/deploy/lattice/target_group_synthesizer.go:426]: Why dont we fix lattice data store at first place? wont it solve bug too?

File: pkg/gateway/model_build_targetgroup.go [hunk 6]
```diff
@@ -172,65 +164,82 @@ func (t *latticeServiceModelBuildTask) buildTargets(ctx context.Context) error {
...
-			K8SServiceName:      t.serviceExport.Name,
-			K8SServiceNamespace: t.serviceExport.Namespace,
-			Protocol:            "HTTP",
-			ProtocolVersion:     vpclattice.TargetGroupProtocolVersionHttp1,
+			Port:                  80,
+			IsServiceImport:       false,
+			IsServiceExport:       true,
+			K8SServiceName:        t.serviceExport.Name,
+			K8SServiceNamespace:   t.serviceExport.Namespace,
+			K8SHTTPRouteName:      "", // When build TG model for serviceExport, the httproutename should be ""
```
Comment (mikhail-aws) [inline pkg/gateway/model_build_targetgroup.go:206]: we should wrap these into different constructor functions, that prevents to pass wrong arguments into struct

File: pkg/gateway/model_build_targetgroup.go [hunk 7]
```diff
@@ -172,65 +164,82 @@ func (t *latticeServiceModelBuildTask) buildTargets(ctx context.Context) error {
...
-		// triggered by serviceexport delete
-		t.Datastore.SetTargetGroupByServiceExport(tgName, false, false)
-	} else {
-		// triggered by serviceexport add
-		t.Datastore.SetTargetGroupByServiceExport(tgName, false, true)
-	}
+	})
+	glog.V(6).Infof("buildTargetGroup, tg[%s], tgSpec%v \n", targetGroupName, tg)
+	// add targetgroup to Datastore local cache for service reconcile to reference
+	// for serviceexport, the httproutename is ""
```
Comment (mikhail-aws) [inline pkg/gateway/model_build_targetgroup.go:214]: please cleanup redundant comments that repeat code below

File: pkg/gateway/model_build_targetgroup.go [hunk 8]
```diff
@@ -172,65 +164,82 @@ func (t *latticeServiceModelBuildTask) buildTargets(ctx context.Context) error {
...
+	t.Datastore.AddTargetGroup(targetGroupName, "", "", "", tg.Spec.Config.IsServiceImport, "")
```
Comment (mikhail-aws) [inline pkg/gateway/model_build_targetgroup.go:215]: we need to refactor datastore that it does not except wrong arguments, especially when we expect some of them being empty strings only

**Comment by liwenwu-amazon** [comment]
I would recommend to have 2 PRs for this, 
* 1st one just to fix [one-line bug](https://github.com/aws/aws-application-networking-k8s/blob/main/pkg/gateway/model_build_targetgroup.go#L186),  where it should not return error if it is delete and k8s service not found
* 2nd one do what ever refactoring you think it is necessary

Right now, it is very hard to see what you are trying to fix. or you are trying to improve

File: pkg/gateway/model_build_targetgroup.go [hunk 9]
```diff
@@ -72,47 +73,38 @@ func (b *targetGroupBuilder) Build(ctx context.Context, srvExport *mcs_api.Servi
...
+	err := t.buildTargetGroupForServiceExport(ctx)
 
 	if err != nil {
 		glog.V(6).Infof("Failed to build TargetGroup when serviceExport buildModel for name %v namespace %v \n", t.serviceExport.Name, t.serviceExport.Namespace)
 		return err
 	}
+	if !t.serviceExport.DeletionTimestamp.IsZero() {
+		//for serviceExport deletion request, we don't need to build targets model
+		return nil
+	}
```
Comment (zijun726911) [inline pkg/gateway/model_build_targetgroup.go:94]: for serviceExport deletion request, we don't need to build targets model because the target_group_manager could handle the deregistering targets 
https://github.com/aws/aws-application-networking-k8s/blob/a8eefa1d4c99c6de030f0a6264b5e1a6b10d1b2e/pkg/deploy/lattice/target_group_manager.go#L201

File: pkg/gateway/model_build_targetgroup.go [hunk 10]
```diff
@@ -260,7 +267,13 @@ func (t *latticeServiceModelBuildTask) buildTargetGroup(ctx context.Context, cli
...
 			if t.route.DeletionTimestamp().IsZero() {
 				// to add
-				t.Datastore.SetTargetGroupByBackendRef(tgName, t.route.Name(), tgSpec.Config.IsServiceImport, true)
+				if *httpBackendRef.Kind() == "Service" {
+					t.Datastore.SetTargetGroupByBackendRef(tgName, t.route.Name(), tgSpec.Config.IsServiceImport, true)
+				} else if *httpBackendRef.Kind() == "ServiceImport" {
+					//When build targetGroup model for service Import byBackendRef
+					t.Datastore.SetTargetGroupByBackendRef(tgName, "", true, true)
+					t.Datastore.SetTargetGroupByBackendRef(tgName, "", false, true)
+				}
```
[the remark this answers is not in the document]
Comment (mikhail-aws) [reply pkg/gateway/model_build_targetgroup.go:278]: I appreciate comments you put around the code, but it should be either explicit in code, or these comments should be in place in source code describing public methods. Comments will be gone with PR, but logic will remain and keep confuse users.

**Comment by zijun726911** [comment]
> I would recommend to have 2 PRs for this,
> 
>     * 1st one just to fix [one-line bug](https://github.com/aws/aws-application-networking-k8s/blob/main/pkg/gateway/model_build_targetgroup.go#L186),  where it should not return error if it is delete and k8s service not found
> 
>     * 2nd one do what ever refactoring you think it is necessary
> 
> 
> Right now, it is very hard to see what you are trying to fix. or you are trying to improve

I once tried to fix the one line fix (or keep diff as little as possible), but it actually didn't work, the reason it don't work is: we show pass the correct LatticeID to the target_group_manage to delete that TG, however I think we didn't assign the correct tgId in cache when we create cache record when we create ServiceExport. I need to fix the way how we add/update/delete latticeStore as well

**Review by mikhail-aws — CHANGES_REQUESTED** [review]
I think we should refactor datastore first, it's hard to follow existing logic. Also please add description to PR page why it's happening. 

As I understand we pass non-empty string to data store that cause datastore return error, and we didnt build model. But when we create datastore do we pass non-empty string too? Why miss-match happens on deletion?

File: pkg/gateway/model_build_targetgroup.go (hunk 9 above)
Comment (zijun726911) [reply pkg/gateway/model_build_targetgroup.go:94]: I think for create/ update serviceExport, the t.serviceExport.DeletionTimestamp.IsZero() == true? So update serviceExport can update the target?

For delete serviceExport t.serviceExport.DeletionTimestamp.IsZero() == false and we be sure need to de-register all targets for that tg and delelete that target group?

**Comment by solmonk** [comment]
Please minimize the refactoring unless it is absolutely necessary. Also create issues instead of TODO comments which makes the diff too big

File: pkg/latticestore/latticestore.go (hunk 1 above)
Comment (solmonk) [reply pkg/latticestore/latticestore.go:405]: please remove these TODO comments and create an issue instead

File: pkg/gateway/model_build_targetgroup.go (hunk 8 above)
Comment (zijun726911) [reply pkg/gateway/model_build_targetgroup.go:215]: Talk with @solmonk , current datastore indeed hard to understand and need refactoring, but we should do it after GRPCRoute and TargetGroupPolicy delivered.  


And @solmonk also suggest to fix code diff as small as possible

What I am think is:
For Scenario 1: stand-alone serviceExport deletion leak tg, that can be fix by few lines change, but I am not 100% sure need to double confirm)
And For scenario 2: non-stand-alone serviceExport deletion leak tg (non-stand-alone means the serviceImport of that serviceExport referenced in a httproute), I think it need to change where we add/update/delete datastore targetGroup cache, only refactoring datastore could make that code easy to follow.

My next step would be try to use a seperate PR to fix Scenario 1 only(keep code diff as small as possible), and keep current PR open until we delivered the GRPCRoute and TargetGroupPolicy 

If quick fix for Scenario 1, I will just go ahead to work on something else

**Comment by mikhail-aws** [comment]
still working on it?

**Comment by zijun726911** [comment]
> still working on it?

Working on another PR to only solve standalone serviceExport deletion leak tg, that one will have smaller diff and more easy to review

**Comment by zijun726911** [comment]
Partial fix for that bug: https://github.com/aws/aws-application-networking-k8s/pull/389

For the other part of fix (ServiceExport's serviceImport is referenced in the route), I prefer to wait until the target group datastore refactoring, maybe from @erikfuller ?
